Documentation/Evolution/StringProcessingAlgorithms.md

```diff
@@ -0,0 +1,402 @@
+        _ regex: R,
+        with replacement: (RegexMatch<R.Match>) throws -> Replacement,
+        subrange: Range<Index>,
+        maxReplacements: Int = .max
+    ) rethrows -> Self where Replacement.Element == Element
+  
+    public func replacing<R: RegexProtocol, Replacement: Collection>(
+        _ regex: R,
+        with replacement: (RegexMatch<R.Match>) throws -> Replacement,
+        maxReplacements: Int = .max
+    ) rethrows -> Self where Replacement.Element == Element
+  
+    public mutating func replace<R: RegexProtocol, Replacement: Collection>(
+        _ regex: R,
+        with replacement: (RegexMatch<R.Match>) throws -> Replacement,
+        maxReplacements: Int = .max
+    ) rethrows where Replacement.Element == Element
+}
+```
+
+#### Split
+
+```swift
+extension Collection where Element: Equatable {
+    public func split<S: Sequence>(by separator: S) -> some Collection<SubSequence>
+        where S.Element == Element
+}
+
+extension BidirectionalCollection where SubSequence == Substring {
+    public func split<R: RegexProtocol>(by separator: R) -> some Collection<Substring>
+}
+```
+
+### `CustomRegexComponent` protocol
+
+The `CustomRegexComponent` protocol inherits from `RegexProtocol` and satisfies its sole requirement. This enables the usage of types that conform to `CustomRegexComponent` in regex builders and `RegexProtocol` algorithms.
+
+```swift
+public protocol CustomRegexComponent: RegexProtocol {
+    /// Match the input string within the specified bounds, beginning at the given index, and return
+    /// the end position (upper bound) of the match and the matched instance.
+    /// - Parameters:
+    ///   - input: The string in which the match is performed.
+    ///   - index: An index of `input` at which to begin matching.
+    ///   - bounds: The bounds in `input` in which the match is performed.
+    /// - Returns: The upper bound where the match terminates and a matched instance, or nil if
+    ///   there isn't a match.
+    func match(
+        _ input: String,
+        startingAt index: String.Index,
+        in bounds: Range<String.Index>
+    ) -> (upperBound: String.Index, match: Match)?
+}
+```
+
+You can conform your custom parser to `CustomRegexComponent`. Conformance is simple: implement the `match` function to return the upper bound of the matched substring, and the type represented by the matched range. It inherits from `RegexProtocol`, so you will be able to use it with all of the string algorithms that take a `RegexProtocol` type.
+
+Here, we use Foundation framework's `FloatingPointFormatStyle<Double>.Currency` as an example. `FloatingPointFormatStyle<Double>.Currency` would conform to `CustomRegexComponent` by implementing the `match` function, with `Match` being a `Double`. It could also add a static function `.localizedCurrency(code:)` as a member of `RegexProtocol`, so you can refer to it as `.localizedCurrency(code:)` in the `Regex` result builder. 
```

Comment:
It really seems like this should be merged into the "proposed solution" sub-section, where the example's fresh in the reader's mind.